src/main/scala/xiangshan/mem/prefetch/Berti.scala

```diff
@@ -0,0 +1,978 @@
+  val thresholdOfL1PF = Constantin.createRecord(_name+"_thresholdOfL1PF", 4)      // ((1 << DtCntWidth) * 0.65).toInt
+  val thresholdOfL2PF = Constantin.createRecord(_name+"_thresholdOfL2PF", 2)      // ((1 << DtCntWidth) * 0.5).toInt
+  val thresholdOfL2PFR = Constantin.createRecord(_name+"_thresholdOfL2PFR", 1)    // ((1 << DtCntWidth) * 0.35).toInt
+  def getPcTag(pc: UInt): UInt = {
+    val res = getPCHash(pc)
+    res(DtPcTagWidth - 1, 0)
+  }
+  def getStatus(conf: UInt): DeltaStatus.Type = {
+    val res = Wire(DeltaStatus())
+    when(conf >= thresholdOfL1PF){
+      res := DeltaStatus.L1_PREF
+    }.elsewhen(conf > thresholdOfL2PF){
+      res := DeltaStatus.L2_PREF
+    }.elsewhen(conf > thresholdOfL2PFR) {
+      res := DeltaStatus.L2_PREF_REPL
+    }.otherwise{
+      res := DeltaStatus.NO_PREF
+    }
+    res
+  }
+
+  /*** built-in class */
+  class DeltaInfo()(implicit p: Parameters) extends BertiBundle {
+    val delta = SInt(DeltaWidth.W)
+    val coverageCnt = UInt(DtCntWidth.W)
+    val status = DeltaStatus()
+
+    def init(): Unit = {
+      delta := 0.S
+      coverageCnt := 0.U
+      status := DeltaStatus.NO_PREF
+    }
+
+    def set(_delta: SInt): Unit = {
+      delta := _delta
+      coverageCnt := 1.U
+    }
+
+    def update(inc: UInt = 1.U): Unit = {
+      coverageCnt := coverageCnt + inc
+    }
+
+    // enter the new cycle
+    // use next to use this record
+    def newCycle(next: UInt = 0.U): Unit = {
+      coverageCnt := 0.U
+      status := getStatus(Mux(next === 0.U, coverageCnt, next))
+    }
+
+    // enter the new status
+    // use next to use this record
+    def newStatus(next: UInt = 0.U): Unit = {
+      status := getStatus(Mux(next === 0.U, coverageCnt, next))
+    }
+  }
+
+  class DeltaEntry()(implicit p: Parameters) extends BertiBundle {
+    val pcTag = UInt(DtPcTagWidth.W)
+    val counter = UInt(DtCntWidth.W)
+    val bestDeltaIdx = UInt(DtDeltaIndexWidth.W)
+    val deltaList = Vec(DtDeltaSize, new DeltaInfo())
+
+    def init(): Unit = {
+      pcTag := 0.U
+      counter := 0.U
+      bestDeltaIdx := 0.U
+      deltaList.map(x => x.init())
+    }
+
+    def setStatus(): Unit = {
+      when(counter >= thresholdOfReset){
+        counter := 0.U
+        deltaList.map(x => x.newCycle())
+      }.elsewhen(counter >= thresholdOfUpdate){
+        deltaList.map(x => x.newStatus())
+      }
+      // when(counter >= thresholdOfUpdate){
+      //   counter := 0.U
+      //   deltaList.map(x => x.newCycle())
+      // }
+    }
+
+    // TODO: use next value to set status
+    // Because that way you don't lose that value from the latest update?
+    // But it's a little hard to record the nextDeltaCnt
+    def setStatus(nextCnt: UInt, nextDeltaCnt: Seq[UInt]): Unit = {
+      when(nextCnt >= thresholdOfReset){
+        counter := 0.U
+        deltaList.zip(nextDeltaCnt).map{case (x, cnt) => x.newCycle(cnt)}
+      }.elsewhen(counter >= thresholdOfUpdate){
```

Comment:
Should we use `nextCnt` here instead of `counter`?

```suggestion
      }.elsewhen(nextCnt >= thresholdOfUpdate){
```